### ISSUE TITLE
`abm()` call modifies global environment

### ISSUE THREAD

**Issue by sashahafner** [AU-BCE-EE/ABM#53]
This needs to be fixed.

![image](https://github.com/user-attachments/assets/1953a1b7-7275-4829-a7b9-9d11bc1de392)

**Comment by sashahafner** [comment]
Here is a note in the help file for `with()`:

```
Note:

     For _interactive_ use this is very effective and nice to read.
     For _programming_ however, i.e., in one's functions, more care is
     needed, and typically one should refrain from using ‘with()’, as,
     e.g., variables in ‘data’ may accidentally override local
     variables, see the reference.
```

**Comment by sashahafner** [comment]
Yup, that's the exact problem.
So is there an alternative other than our old approach of manual extraction to put all `pars` elements in the `rates()` environment?

**Comment by sashahafner** [comment]
That's 110187652e6b6c4f4fd5368ef8c7ccd5496dc005

**Comment by sashahafner** [comment]
I know the examples for the deSolve package use `with()`.
Interesting though that the paper on it doesn't even seem to include any examples where `parms` is used!
https://journal.r-project.org/articles/RJ-2010-013/
I checked in case there was some discussion on this `with()` issue.

**Comment by sashahafner** [comment]
Well, I'm confused!
In 84f1be869fb3ab18bd6ee16dd21694e0bf17173c I replaced `with()` with `assign()` and still see these variables showing up in the main environment.
Anyway I guess it should have been obvious that `with(parms)` wasn't the problem, because the variables showing up exclude a lot of elements in pars.

**Comment by sashahafner** [comment]
I see by commenting out that it is `abm_variable()` that adds them all:

![image](https://github.com/user-attachments/assets/e4500eb9-655a-4b09-bfdb-593b96108321)

**Comment by sashahafner** [comment]
Moving around an early return call shows this `lsoda` line causes the problem:

![image](https://github.com/user-attachments/assets/ea8a8441-f050-4cf0-ae2b-0869a7f74e28)

Why?

**Comment by sashahafner** [comment]
Here it is in `rates()`!
![image](https://github.com/user-attachments/assets/62ccfa35-40a5-4029-a58f-3f902f90f4d4)

**Comment by sashahafner** [comment]
Fixed in a389ea661be45a9c756537bd929ec6f857633802.
Do we really want to load all these elements into the `rates()` environment though?
I guess it works OK, and saves us having to figure out which need the `y['']` part.
